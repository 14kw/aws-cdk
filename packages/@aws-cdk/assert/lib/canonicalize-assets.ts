import * as cxapi from '@aws-cdk/cx-api';
import * as path from 'path';

/**
 * Reduce template to a normal form where asset references have been normalized
 *
 * This makes it possible to compare templates if all that's different between
 * them is the hashes of the asset values.
 *
 * Currently only handles parameterized assets, but can (and should)
 * be adapted to handle convention-mode assets as well when we start using
 * more of those.
 */
export function canonicalizeStackArtifact(artifact: cxapi.CloudFormationStackArtifact): any {
  // Find assets via parameters
  const stringSubstitutions = new Array<[RegExp, string]>();

  const assets = findAssets(artifact);

  assets.forEach((asset, i) => {
    const ix = i + 1;
    switch (asset.type) {
      case 'legacy-file':
        // Full parameter reference
        stringSubstitutions.push([new RegExp(asset.s3BucketParameter), `Asset${ix}S3Bucket`]);
        stringSubstitutions.push([new RegExp(asset.s3KeyParameter), `Asset${ix}S3VersionKey`]);
        stringSubstitutions.push([new RegExp(asset.artifactHashParameter), `Asset${ix}ArtifactHash`]);

        // Substring asset hash reference
        stringSubstitutions.push([ new RegExp(asset.sourceHash), `Asset${ix}Hash`]);
    }
  });

  const substitutedTemplate = substitute(artifact.template);

  // We have the template, now turn it into a CloudFormationStackArtifact to return
  // Write it into a new Cloud Assembly in a subdirectory of the existing assembly so everything gets
  // cleaned up together.
  const builder = new cxapi.CloudAssemblyBuilder(path.join(artifact.assembly.directory, `${artifact.templateFile}.canonicalized`));
  builder.writeStack(artifact.id, {
    template: substitutedTemplate,
  });
  return builder.buildAssembly().getStackByName(artifact.id);

  function substitute(what: any): any {
    if (Array.isArray(what)) {
      return what.map(substitute);
    }

    if (typeof what === 'object' && what !== null) {
      const ret: any = {};
      for (const [k, v] of Object.entries(what)) {
        ret[stringSub(k)] = substitute(v);
      }
      return ret;
    }

    if (typeof what === 'string') {
      return stringSub(what);
    }

    return what;
  }

  function stringSub(x: string) {
    for (const [re, replacement] of stringSubstitutions) {
      x = x.replace(re, replacement);
    }
    return x;
  }
}

type AssetInfo = (
  { type: 'legacy-file'; sourceHash: string; s3BucketParameter: string; s3KeyParameter: string; artifactHashParameter: string; }
);

function findAssets(artifact: cxapi.CloudFormationStackArtifact) {
  const ret: AssetInfo[] = [];
  for (const asset of artifact.assets) {
    switch (asset.packaging) {
      case 'zip':
      case 'file':
        ret.push({
          type: 'legacy-file',
          sourceHash: asset.sourceHash,
          s3KeyParameter: asset.s3KeyParameter,
          s3BucketParameter: asset.s3BucketParameter,
          artifactHashParameter: asset.artifactHashParameter,
        });
        continue;
    }

    throw new Error(`Cannot canonicalize legacy asset of type ${asset.packaging} yet`);
  }
  return ret;
}
