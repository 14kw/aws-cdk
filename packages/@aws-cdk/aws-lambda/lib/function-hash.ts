import * as crypto from 'crypto';
import { CfnResource, Stack } from '@aws-cdk/core';
import { Function } from './function';
import { sortedStringify } from './private/sort';

export const ENV_SALT = '4b2bff42-e517-41d1-9b5f-92aacf41842b';

/*
 * The list of properties found in CfnFunction (or AWS::Lambda::Function).
 * They are classified as "locked" to a Function Version or not.
 * When a property is locked, any change to that property will not take effect on previously created Versions.
 * Instead, a new Version must be generated for the change to take effect.
 * Similarly, if a property that's not locked to a Version is modified, a new Version
 * must not be generated.
 *
 * Adding a new property to this list - If the property is part of the UpdateFunctionConfiguration
 * API, then it must be classified as true, otherwise false.
 * See https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionConfiguration.html
 */
const VERSION_PROPS_CLASSIFICATION: { [key: string]: boolean } = {
  // locked to the version
  Code: true,
  DeadLetterConfig: true,
  Description: true,
  Environment: true,
  FileSystemConfigs: true,
  FunctionName: true,
  Handler: true,
  KmsKeyArn: true,
  Layers: true,
  MemorySize: true,
  Role: true,
  Runtime: true,
  Timeout: true,
  TracingConfig: true,
  VpcConfig: true,

  // not locked to the version
  CodeSigningConfigArn: false,
  ReservedConcurrentExecutions: false,
  Tags: false,
};

export interface CalculateFunctionHashOptions {
  hashAlgorithm: 'v1' | 'v2';
  function: Function;
  additionalVersionPropsClassification?: { [key: string]: boolean };
}

export function calculateFunctionHash(options: CalculateFunctionHashOptions) {
  const stack = Stack.of(options.function);

  const functionResource = options.function.node.defaultChild as CfnResource;

  // render the cloudformation resource from this function
  const config = stack.resolve((functionResource as any)._toCloudFormation());
  const hash = crypto.createHash('md5');

  if (options.hashAlgorithm === 'v2') {
    hash.update(sortedStringify(usefulKeys(config, options.additionalVersionPropsClassification)));
  } else {
    hash.update(JSON.stringify(config));
  }

  return hash.digest('hex');
}

function usefulKeys(cfnObject: any, additionalClassification?: { [key: string]: boolean }): any {
  const keys = Object.keys(cfnObject.Resources);
  if (keys.length !== 1) {
    throw new Error(`Expected one rendered CloudFormation resource but found ${keys.length}`);
  }
  const props = cfnObject.Resources[keys[0]].Properties;

  const classification = {
    ...VERSION_PROPS_CLASSIFICATION,
    ...additionalClassification,
  };

  const unclassified = Object.keys(props).filter(p => !Object.keys(classification).includes(p));
  if (unclassified.length > 0) {
    throw new Error(`Keys [${unclassified}] are unclassified on whether they are version locked.`
      + 'Use "additionalVersionPropsClassification" property to classify them.');
  }
  const notVersionLocked = Object.entries(classification).filter(e => !e[1]).map(e => e[0]);
  notVersionLocked.forEach(p => delete props[p]);
  return props;
}

export function trimFromStart(s: string, maxLength: number) {
  const desiredLength = Math.min(maxLength, s.length);
  const newStart = s.length - desiredLength;
  return s.substring(newStart);
}