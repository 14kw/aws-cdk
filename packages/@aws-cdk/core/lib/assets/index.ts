export * from './file-asset';
export * from './image-asset';
export * from './common';
export * from './asset-staging';
export * from './bundling';