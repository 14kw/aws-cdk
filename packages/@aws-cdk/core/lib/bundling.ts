import { spawnSync, SpawnSyncOptions } from 'child_process';
import * as crypto from 'crypto';
import { FileSystem } from './fs';

/**
 * Bundling options
 *
 * @experimental
 */
export interface BundlingOptions {
  /**
   * The Docker image where the command will run.
   */
  readonly image: BundlingDockerImage;

  /**
   * The command to run in the Docker container.
   *
   * @example ['npm', 'install']
   *
   * @see https://docs.docker.com/engine/reference/run/
   *
   * @default - run the command defined in the image
   */
  readonly command?: string[];

  /**
   * Additional Docker volumes to mount.
   *
   * @default - no additional volumes are mounted
   */
  readonly volumes?: DockerVolume[];

  /**
   * The environment variables to pass to the Docker container.
   *
   * @default - no environment variables.
   */
  readonly environment?: { [key: string]: string; };

  /**
   * Working directory inside the Docker container.
   *
   * @default /asset-input
   */
  readonly workingDirectory?: string;

  /**
   * The user to use when running the Docker container.
   *
   *   user | user:group | uid | uid:gid | user:gid | uid:group
   *
   * @see https://docs.docker.com/engine/reference/run/#user
   *
   * @default - uid:gid of the current user or 1000:1000 on Windows
   */
  readonly user?: string;

  /**
   * Local bundling provider.
   *
   * The provider implements a method `tryBundle()` which should return `true`
   * if local bundling was performed. If `false` is returned, docker bundling
   * will be done.
   *
   * @default - bundling will only be performed in a Docker container
   *
   * @experimental
   */
  readonly local?: ILocalBundling;
}

/**
 * Local bundling
 *
 * @experimental
 */
export interface ILocalBundling {
  /**
   * This method is called before attempting docker bundling to allow the
   * bundler to be executed locally. If the local bundler exists, and bundling
   * was performed locally, return `true`. Otherwise, return `false`.
   *
   * @param outputDir the directory where the bundled asset should be output
   * @param options bundling options for this asset
   */
  tryBundle(outputDir: string, options: BundlingOptions): boolean;
}

/**
 * Options for starting a container.
 */
export interface DockerStartOptions {

  /**
   * Attach STDOUT/STDERR and forward signals.
   *
   * @default false
   */
  readonly attach?: boolean;
}

/**
 * A Docker image used for asset bundling
 */
export class BundlingDockerImage {
  /**
   * Reference an image on DockerHub or another online registry.
   *
   * @param image the image name
   */
  public static fromRegistry(image: string) {
    return new BundlingDockerImage(image);
  }

  /**
   * Reference an image that's built directly from sources on disk.
   *
   * @param path The path to the directory containing the Docker file
   * @param options Docker build options
   */
  public static fromAsset(path: string, options: DockerBuildOptions = {}) {
    const buildArgs = options.buildArgs || {};

    // Image tag derived from path and build options
    const tagHash = crypto.createHash('sha256').update(JSON.stringify({
      path,
      ...options,
    })).digest('hex');
    const tag = `cdk-${tagHash}`;

    const dockerArgs: string[] = [
      'build', '-t', tag,
      ...(options.file ? ['-f', options.file] : []),
      ...flatten(Object.entries(buildArgs).map(([k, v]) => ['--build-arg', `${k}=${v}`])),
      path,
    ];

    dockerExec(dockerArgs);

    // Fingerprints the directory containing the Dockerfile we're building and
    // differentiates the fingerprint based on build arguments. We do this so
    // we can provide a stable image hash. Otherwise, the image ID will be
    // different every time the Docker layer cache is cleared, due primarily to
    // timestamps.
    const hash = FileSystem.fingerprint(path, { extraHash: JSON.stringify(options) });
    return new BundlingDockerImage(tag, hash);
  }

  /** @param image The Docker image */
  private constructor(public readonly image: string, private readonly _imageHash?: string) {}

  /**
   * Provides a stable representation of this image for JSON serialization.
   *
   * @return The overridden image name if set or image hash name in that order
   */
  public toJSON() {
    return this._imageHash ?? this.image;
  }

  /**
   * Runs a Docker image
   */
  public run(options: DockerRunOptions = {}) {
    const container = this.create(options);
    try {
      const volumes = options.volumes || [];
      for (const v of volumes) {
        container.copyTo(`${v.hostPath}/.`, v.containerPath);
      }
      container.start();
      for (const v of volumes) {
        container.copyFrom(`${v.containerPath}/.`, v.hostPath);
      }
    } finally {
      container.remove();
    }
  }

  /**
   * Copies a file or directory out of the Docker image to the local filesystem
   */
  public cp(imagePath: string, outputPath: string) {
    const container = this.create();
    try {
      container.copyFrom(imagePath, outputPath);
    } finally {
      container.remove();
    }
  }

  /**
   * Create an container instance from this image.
   * Note that the container is only created, not started. To start the container, call `.start()`
   * on the return value.
   *
   * @see BundlingDockerContainer
   * @param options create options.
   */
  public create(options: DockerCreateOptions = {}): BundlingDockerContainer {

    const environment = options?.environment || {};
    const command = options?.command || [];

    const dockerArgs: string[] = [
      'create',
      ...options?.user
        ? ['-u', options.user]
        : [],
      ...flatten(Object.entries(environment).map(([k, v]) => ['--env', `${k}=${v}`])),
      ...options?.workingDirectory
        ? ['-w', options.workingDirectory]
        : [],
      this.image,
      ...command,
    ];

    const { stdout } = dockerExec(dockerArgs, {
      stdio: [
        'ignore', // ignore stdin
        'pipe', // pipe stdout
        process.stdout, // redirect stderr to stdout
      ],
    });
    const match = stdout.toString().match(/([0-9a-f]{16,})/);
    if (!match) {
      throw new Error('Failed to extract container ID from Docker create output');
    }
    return new BundlingDockerContainer(match[1]);
  }
}

/**
 * A Docker volume
 */
export interface DockerVolume {
  /**
   * The path to the file or directory on the host machine
   */
  readonly hostPath: string;

  /**
   * The path where the file or directory is mounted in the container
   */
  readonly containerPath: string;

  /**
   * Mount consistency. Only applicable for macOS
   *
   * @default DockerConsistency.DELEGATED
   * @see https://docs.docker.com/storage/bind-mounts/#configure-mount-consistency-for-macos
   */
  readonly consistency?: DockerVolumeConsistency;
}

/**
 * Supported Docker volume consistency types. Only valid on macOS due to the way file storage works on Mac
 */
export enum DockerVolumeConsistency {
  /**
   * Read/write operations inside the Docker container are applied immediately on the mounted host machine volumes
   */
  CONSISTENT = 'consistent',
  /**
   * Read/write operations on mounted Docker volumes are first written inside the container and then synchronized to the host machine
   */
  DELEGATED = 'delegated',
  /**
   * Read/write operations on mounted Docker volumes are first applied on the host machine and then synchronized to the container
   */
  CACHED = 'cached',
}

/**
 * Docker create options
 */
export interface DockerCreateOptions {

  /**
   * The command to run in the container.
   *
   * @default - run the command defined in the image
   */
  readonly command?: string[];

  /**
   * The environment variables to pass to the container.
   *
   * @default - no environment variables.
   */
  readonly environment?: { [key: string]: string; };

  /**
   * Working directory inside the container.
   *
   * @default - image default
   */
  readonly workingDirectory?: string;

  /**
   * The user to use when running the container.
   *
   * @default - root or image default
   */
  readonly user?: string;

}

/**
 * Docker run options
 */
export interface DockerRunOptions extends DockerCreateOptions {

  /**
   * Docker volumes to mount.
   *
   * @default - no volumes are mounted
   */
  readonly volumes?: DockerVolume[];

}

/**
 * Docker build options
 */
export interface DockerBuildOptions {
  /**
   * Build args
   *
   * @default - no build args
   */
  readonly buildArgs?: { [key: string]: string };

  /**
   * Name of the Dockerfile
   *
   * @default - The Dockerfile immediately within the build context path
   */
  readonly file?: string;
}

function flatten(x: string[][]) {
  return Array.prototype.concat([], ...x);
}

function dockerExec(args: string[], options?: SpawnSyncOptions) {
  const prog = process.env.CDK_DOCKER ?? 'docker';
  const proc = spawnSync(prog, args, options ?? {
    stdio: [ // show Docker output
      'ignore', // ignore stdio
      process.stderr, // redirect stdout to stderr
      'inherit', // inherit stderr
    ],
  });

  if (proc.error) {
    throw proc.error;
  }

  if (proc.status !== 0) {
    if (proc.stdout || proc.stderr) {
      throw new Error(`[Status ${proc.status}] stdout: ${proc.stdout?.toString().trim()}\n\n\nstderr: ${proc.stderr?.toString().trim()}`);
    }
    throw new Error(`${prog} exited with status ${proc.status}`);
  }

  return proc;
}

/**
 * A docker container used for asset bundling.
 */
export class BundlingDockerContainer {

  constructor(
    /**
     * The container id.
     */
    public readonly id: string,
  ) {}

  /**
   * Copy files/directories from the host to the container.
   *
   * @param hostPath source path on the host.
   * @param containerPath destination path in the container.
   */
  public copyTo(hostPath: string, containerPath: string) {
    return this.cp(hostPath, `${this.id}:${containerPath}`);
  }

  /**
   * Copy files/directories from the container to the host.
   *
   * @param containerPath source path in the container.
   * @param hostPath destination path on the host.
   */
  public copyFrom(containerPath: string, hostPath: string) {
    return this.cp(`${this.id}:${containerPath}`, hostPath);
  }

  /**
   * Start the container.
   *
   * @param options start options.
   */
  public start(options: DockerStartOptions = {}) {
    const args = ['start'];
    if (options.attach ?? false) {
      args.push('-a');
    }
    dockerExec([...args, this.id]);
  }

  /**
   * Remove the container and all associated volumes.
   */
  public remove() {
    dockerExec(['rm', '-vf', this.id]);
  }

  private cp(src: string, dst: string) {
    try {
      dockerExec(['cp', src, dst]);
    } catch (err) {
      throw new Error(`Failed to copy files from ${src} to ${dst}: ${err}`);
    }
  }
}
