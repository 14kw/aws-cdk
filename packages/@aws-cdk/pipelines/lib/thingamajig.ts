import * as codepipeline from '@aws-cdk/aws-codepipeline';
import { Stage } from '@aws-cdk/core';
import * as cxapi from '@aws-cdk/cx-api';
import { Construct } from 'constructs';
import { actionsFromStackArtifact } from './actions/_action_maker';
import { topologicalSort } from './private/toposort';

export interface IPipelineBuilder {
  scope: Construct;
  cloudAssemblyArtifact: codepipeline.Artifact;
  addAction(action: codepipeline.IAction): void;
  addArtifactsFor(stackArtifact: cxapi.CloudFormationStackArtifact): void;
}

export interface IPipelineable {
  addToPipeline(builder: IPipelineBuilder): void;
}

export interface AppContext {
}

export interface ChangeSetsContext {
}
export interface IAppApprover {
  approveApp(appContext: AppContext, builder: IPipelineBuilder): void;
}

export interface IChangeSetApprover {
  approveChangeSets(changeSetsContext: ChangeSetsContext, builder: IPipelineBuilder): void;
}
export interface DeployStageProps {
  readonly appStage: Stage;
  readonly changeSetApprover?: IChangeSetApprover;
  readonly approvalSteps?: IAppApprover[];
}

export class DeployStage implements IPipelineable {
  private readonly asm: cxapi.CloudAssembly;

  constructor(private readonly props: DeployStageProps) {
    this.asm = props.appStage.synth();

    if (this.asm.stacks.length === 0) {
      // If we don't check here, a more puzzling "stage contains no actions"
      // error will be thrown come deployment time.
      throw new Error(`The given Stage construct ('${props.appStage.node.path}') should contain at least one Stack`);
    }
  }

  public addToPipeline(builder: IPipelineBuilder): void {
    const sortedTranches = topologicalSort(this.asm.stacks,
      stack => stack.id,
      stack => stack.dependencies.map(d => d.id));

    for (const stacks of sortedTranches) {
      // These don't have a dependency on each other, so can all be added in parallel

      // Get all assets manifests and add the assets in 'em to the asset publishing stage.
      for (const stackArtifact of stacks) {
        builder.addArtifactsFor(stackArtifact);
      }

      const actionSets = stacks.map(artifact => actionsFromStackArtifact(builder.scope, artifact, {
        cloudAssemblyInput: builder.cloudAssemblyArtifact,
      }));

      for (const actionSet of actionSets) {
        builder.addAction(actionSet.prepareAction);
      }

      this.props.changeSetApprover?.approveChangeSets({
      }, builder);

      for (const actionSet of actionSets) {
        builder.addAction(actionSet.executeAction);
      }
    }

    for (const approval of this.props.approvalSteps ?? []) {
      approval.approveApp({
      }, builder);
    }
  }
}