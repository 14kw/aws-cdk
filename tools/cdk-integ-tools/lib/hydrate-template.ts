import * as cxschema from '@aws-cdk/cloud-assembly-schema';
import * as cxapi from '@aws-cdk/cx-api';

/**
 * Hydrate a template into a StackArtifact
 *
 * Detects asset references in the template and add them to metadata. This is
 * necessary to normalize the assets out again afterwards in the comparison
 * (because canonicalization function goes from StackArtifact -> StackArtifact).
 */
export function hydrateTemplateToStackArtifact(template: any): cxapi.CloudFormationStackArtifact {
  const fileParamRe = /^AssetParameters([a-zA-Z0-9]{64})(S3Bucket|S3VersionKey|ArtifactHash)([a-zA-Z0-9]{8})$/;

  const fileAssets: Record<string, Writeable<cxschema.FileAssetMetadataEntry>> = {};

  for (const paramName of Object.keys(template?.Parameters || {})) {
    const m = fileParamRe.exec(paramName);
    if (!m) { continue; }

    const key = m[1];
    if (!fileAssets[key]) {
      fileAssets[key] = {
        packaging: 'file',
        sourceHash: m[1],
        artifactHashParameter: '',
        s3KeyParameter: '',
        s3BucketParameter: '',
      };
    }
    const fileMeta = fileAssets[key];

    switch (m[2]) {
      case 'S3Bucket':
        fileMeta.s3BucketParameter = paramName;
        break;
      case 'S3VersionKey':
        fileMeta.s3KeyParameter = paramName;
        break;
      case 'ArtifactHash':
        fileMeta.artifactHashParameter = paramName;
        break;
    }
  }

  const metadata: Record<string, cxapi.MetadataEntry[]> = {};
  for (const [hash, data] of Object.entries(fileAssets)) {
    metadata[`/asset${hash}`] = [
      {
        type: cxschema.ArtifactMetadataEntryType.ASSET,
        data: data as any, // Trust me I know what I'm doing
      },
    ];
  }

  // Return as a StackArtifact (need to write a temporary assembly here)
  const builder = new cxapi.CloudAssemblyBuilder();
  builder.writeStack('Expectation', {
    template,
    metadata,
  });
  return builder.buildAssembly().getStackByName('Expectation');
}

/**
 * Unfortunately there is no way to preserve optionality in this mapped type, so mark everything optional otherwise code becomes annoying
 */
type Writeable<T> = { -readonly [P in keyof T]?: T[P] };
