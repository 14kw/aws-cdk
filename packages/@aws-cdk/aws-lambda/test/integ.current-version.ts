import * as path from 'path';
import { App, CfnOutput, RemovalPolicy, Stack } from '@aws-cdk/core';
import * as lambda from '../lib';

class TestStack extends Stack {
  constructor(scope: App, id: string) {
    super(scope, id);

    const handler = new lambda.Function(this, 'MyLambda', {
      code: lambda.Code.fromAsset(path.join(__dirname, 'layer-code')),
      handler: 'index.main',
      runtime: lambda.Runtime.PYTHON_3_8,
      currentVersionOptions: {
        removalPolicy: RemovalPolicy.RETAIN,
        retryAttempts: 1,
      },
    });

    handler.currentVersion.addAlias('live');

    const edgeLambda = new lambda.Function(this, 'MyLambda@Edge', {
      code: lambda.Code.fromAsset(path.join(__dirname, 'layer-code')),
      handler: 'index.main',
      runtime: lambda.Runtime.PYTHON_3_8,
      currentVersionOptions: {
        removalPolicy: RemovalPolicy.RETAIN,
        retryAttempts: 1,
      },
    });

    new CfnOutput(this, 'EdgeArn', {
      value: edgeLambda.currentVersion.edgeArn,
    });
  }
}

const app = new App();

new TestStack(app, 'lambda-test-current-version');

app.synth();
