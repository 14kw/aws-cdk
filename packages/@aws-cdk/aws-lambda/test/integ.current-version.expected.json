{
  "Resources": {
    "MyLambdaServiceRole4539ECB6": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "MyLambdaCCE802FB": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "AssetParameters5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3cS3BucketE8900D83"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3cS3VersionKeyFD066EB8"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "AssetParameters5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3cS3VersionKeyFD066EB8"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.main",
        "Role": {
          "Fn::GetAtt": [
            "MyLambdaServiceRole4539ECB6",
            "Arn"
          ]
        },
        "Runtime": "python3.8"
      },
      "DependsOn": [
        "MyLambdaServiceRole4539ECB6"
      ]
    },
    "MyLambdaCurrentVersionE7A382CC16bb4a7abf5d771ad2c2be359290ccde": {
      "Type": "AWS::Lambda::Version",
      "Properties": {
        "FunctionName": {
          "Ref": "MyLambdaCCE802FB"
        }
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "MyLambdaCurrentVersionEventInvokeConfigD120DC68": {
      "Type": "AWS::Lambda::EventInvokeConfig",
      "Properties": {
        "FunctionName": {
          "Ref": "MyLambdaCCE802FB"
        },
        "Qualifier": {
          "Fn::GetAtt": [
            "MyLambdaCurrentVersionE7A382CC16bb4a7abf5d771ad2c2be359290ccde",
            "Version"
          ]
        },
        "MaximumRetryAttempts": 1
      }
    },
    "MyLambdaCurrentVersionAliaslive9151E913": {
      "Type": "AWS::Lambda::Alias",
      "Properties": {
        "FunctionName": {
          "Ref": "MyLambdaCCE802FB"
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "MyLambdaCurrentVersionE7A382CC16bb4a7abf5d771ad2c2be359290ccde",
            "Version"
          ]
        },
        "Name": "live"
      }
    }
  },
  "Parameters": {
    "AssetParameters5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3cS3BucketE8900D83": {
      "Type": "String",
      "Description": "S3 bucket for asset \"5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3c\""
    },
    "AssetParameters5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3cS3VersionKeyFD066EB8": {
      "Type": "String",
      "Description": "S3 key for asset version \"5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3c\""
    },
    "AssetParameters5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3cArtifactHashD5562C20": {
      "Type": "String",
      "Description": "Artifact hash for asset \"5bfcc5e85793015838d125d6e30f1a931da22ebe768a9bad5cd1b17fd6bd0a3c\""
    }
  }
}